### PR TITLE
gui2: indicate that unit cannot advance

### LINKED ISSUE #1244: When no advancement nor AMLA is defined the xp needed should be infinite and showed in the ui as a "-" (GNA #17262)
**[Original submission](https://gna.org/bugs/index.php?17262)** by iwontbecreative on 2010-12-07

When a unit doesn't have any advancement and doesn't define any kind of AMLA, the xp should be unlimited. (Because the unit will never level) and showed in the ui as "-" so people can understand it.

 You can see an example of this bug in the tutorial: The fighter/fighteress needs 28 xp to "level" and his xp bar is light blue. Both are confusing because the unit won't level (Will show e.g 34/28 instead) and that the light blue is for units with an advancement.

(Reproduced on Linux)
Release: trunk. (rev47854)
Priority: 5 - Normal
Severity: 1 - Wish

### PR DESCRIPTION
**Pull request by hryniuk**
Fixes #1244

### CODE REVIEW

**Comment by hryniuk** [comment]
I hope I've change all places with an experience info visible to a player. I've left the exact experience value in the game inspector (under `:inspect` command; `src/gui/dialogs/gamestate_inspector.cpp`), as it's debug info.

File: src/gui/dialogs/unit_list.cpp [hunk 1]
```diff
@@ -119,8 +119,13 @@ void unit_list::pre_show(window& window)
 		row_data.emplace("unit_level", column);
 
 		std::stringstream exp_str;
-		exp_str << font::span_color(unit->xp_color()) << unit->experience() << "/"
-		        << (unit->can_advance() ? std::to_string(unit->max_experience()) : font::unicode_en_dash) << "</span>";
+		exp_str << font::span_color(unit->xp_color());
+		if (unit->can_advance()) {
```
Comment (jyrkive) [inline src/gui/dialogs/unit_list.cpp]: Nitpicking: Wesnoth coding style doesn't have a space after the `if` keyword.

File: src/gui/dialogs/unit_recall.cpp [hunk 2]
```diff
@@ -201,8 +201,13 @@ void unit_recall::pre_show(window& window)
 		row_data.emplace("unit_level", column);
 
 		std::stringstream exp_str;
-		exp_str << font::span_color(unit->xp_color()) << unit->experience() << "/"
-		        << (unit->can_advance() ? std::to_string(unit->max_experience()) : font::unicode_en_dash) << "</span>";
+		exp_str << font::span_color(unit->xp_color());
+		if (unit->can_advance()) {
```
Comment (jyrkive) [inline src/gui/dialogs/unit_recall.cpp]: Space.

File: src/gui/widgets/unit_preview_pane.cpp [hunk 3]
```diff
@@ -457,23 +457,28 @@ void unit_preview_pane::set_displayed_unit(const unit& u)
 		str << font::span_color(u.hp_color())
 			<< _("HP: ") << u.hitpoints() << "/" << u.max_hitpoints() << "</span>" << "\n";
 
-		str << font::span_color(u.xp_color())
-			<< _("XP: ") << u.experience() << "/" << u.max_experience() << "</span>";
+		str << font::span_color(u.xp_color()) << _("XP: ");
+		if (u.can_advance()) {
```
Comment (jyrkive) [inline src/gui/widgets/unit_preview_pane.cpp]: Space.

File: src/gui/widgets/unit_preview_pane.cpp [hunk 4]
```diff
@@ -457,23 +457,28 @@ void unit_preview_pane::set_displayed_unit(const unit& u)
...
+			str << font::unicode_en_dash;
+		}
 
 		label_details_->set_label(str.str());
 		label_details_->set_use_markup(true);
 	}
 
 	if(tree_details_) {
 		tree_details_->clear();
+		const std::string unit_xp = (u.can_advance() ? (formatter() << u.experience() << "/" << u.max_experience()).str() : font::unicode_en_dash);
```
Comment (jyrkive) [inline src/gui/widgets/unit_preview_pane.cpp]: You don't need the outmost parentheses.

File: src/reports.cpp [hunk 5]
```diff
@@ -463,8 +463,12 @@ static config unit_xp(const unit* u)
 {
 	if (!u) return config();
 	std::ostringstream str, tooltip;
-	str << span_color(u->xp_color()) << u->experience()
-		<< '/' << u->max_experience() << naps;
+	if (u->can_advance()) {
```
Comment (jyrkive) [inline src/reports.cpp]: Space.

File: src/reports.cpp [hunk 6]
```diff
@@ -463,8 +463,12 @@ static config unit_xp(const unit* u)
...
+		str << span_color(u->xp_color()) << u->experience() << '/' << u->max_experience() << naps;
+	} else {
+		str << span_color(u->xp_color()) << font::unicode_en_dash  << naps;
```
Comment (jyrkive) [inline src/reports.cpp]: You have two spaces before the last `<<`

File: src/gui/widgets/unit_preview_pane.cpp [hunk 7]
```diff
@@ -457,23 +457,28 @@ void unit_preview_pane::set_displayed_unit(const unit& u)
...
 			<< _("HP: ") << u.hitpoints() << "/" << u.max_hitpoints() << "</span>" << "\n";
 
-		str << font::span_color(u.xp_color())
-			<< _("XP: ") << u.experience() << "/" << u.max_experience() << "</span>";
+		str << font::span_color(u.xp_color()) << _("XP: ");
+		if(u.can_advance()) {
+			str << u.experience() << "/" << u.max_experience();
+		} else {
+			str << font::unicode_en_dash;
+		}
```
[the remark this answers is not in the document]
Comment (hryniuk) [reply src/gui/widgets/unit_preview_pane.cpp]: Oh, sorry, missed that one.

**Comment by jyrkive** [comment]
Merged. Thanks for the pull request! :)

**Comment by hryniuk** [comment]
Great :).

**Pull request merged**